File: src/main/java/scraping/ScrapeSection.java
package scraping;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import nyu.SubjectCode;
import nyu.Term;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scraping.models.SectionAttribute;
import scraping.parse.ParseCatalog;
import scraping.parse.ParseSection;
import scraping.query.QueryCatalog;
import scraping.query.QuerySection;

public class ScrapeSection {
  private static Logger logger =
      LoggerFactory.getLogger("scraping.ScrapeCatalog");

  /**
   * Scraping the catalogs from Albert Mobile given multiple subjecs
   * @param term The term for which we should be scraping
   * @param subjectCodes The subjects for which we should be scraping
   * @return Sequence of List of Courses
   */
  public static Stream<SectionAttribute>
  scrapeFromSection(Term term, List<SubjectCode> subjectCodes,
                    Integer batchSize, Integer batchSizeSections) {
    return QuerySection
        .querySections(
            term,
            ParseCatalog.parseRegistrationNumber(
                QueryCatalog.queryCatalog(term, subjectCodes, batchSize)
                    .collect(Collectors.toList())
                    .toString()),
            batchSizeSections)
        .map(data -> ParseSection.parse(data));
  }

  public static SectionAttribute scrapeFromSection(Term term,
                                                   int registrationNumber) {
    return ParseSection.parse(
        QuerySection.querySection(term, registrationNumber));
  }
}

File: src/main/kotlin/services/scrape_section.kt
package services

import nyu.SubjectCode
import nyu.Term
import scraping.models.SectionAttribute
import mu.KotlinLogging
import java.util.stream.Collectors
import java.util.stream.Stream

private val scraperLogger = KotlinLogging.logger("services.scrape_section")

/**
 * Scraping the catalogs from Albert Mobile given one registration number
 * @param logger The logger to log to during execution of this service
 * @param term The term for which we should be scraping
 * @param subjectCodes The subject for which we should be scraping
 * @return List of courses
 */
fun scrapeFromSection(term: Term, registrationNumber: Int): SectionAttribute {
    return querySection(term, registrationNumber).let { rawData ->
        ParseSection.parse(rawData)
    }
}

/**
 * Scraping the catalogs' sections from Albert Mobile given multiple subject codes
 * @param term The term for which we should be scraping
 * @param subjectCodes The subject for which we should be scraping
 * @return List of courses
 */
fun scrapeFromAllCatalogSection(term: Term, subjectCodes: List<SubjectCode>,
                                batchSize: Int? = null): Stream<SectionAttribute> {
    return querySections(term,
            ParseCatalog.parseRegistrationNumber(
                    queryCatalog(term, subjectCodes).collect(Collectors.toList()).toString()), batchSize)
            .asIterable().map { rawData ->
                ParseSection.parse(rawData)
            }.stream()

}